### PR TITLE
update to Lmod 7(.5.10) (HPC-6313)

### PR DESCRIPTION
**Pull request by boegel**
Still WIP because testing revealed two problems that should be fixed somehow:

* https://github.com/TACC/Lmod/issues/289 => fixed by included patch (changed behaviour in Lmod 7.x that doesn't fit our needs)
* https://github.com/TACC/Lmod/issues/290 => fixed in Lmod 7.5.5

(replacement for #21)

### CODE REVIEW

**Comment by boegel** [comment]
@stdweird Good to go imho, integration tests pass:

```
[15:10:45] vsc40023@node2727:~/vsc-testing/module $ ml --version

Modules based on Lua: Version 7.5.10  2017-07-05 14:04 -05:00
    by Robert McLay mclay@tacc.utexas.edu

[15:11:38] vsc40023@node2727:~/vsc-testing/module $ rpm -qa | grep Lmod
Lmod-7.5.10-4.ug.el7.centos.noarch
[15:11:42] vsc40023@node2727:~/vsc-testing/module $ ./run_all_tests.sh
Lmod-7.5.10-4.ug.el7.centos.noarch
vsc-cluster-modules-0.21-1.noarch
vsc-cluster-modules-tier2-0.21-1.noarch

> module --version

Modules based on Lua: Version 7.5.10  2017-07-05 14:04 -05:00
$MODULEPATH: /apps/gent/CO7/haswell-ib/modules/all:/etc/modulefiles/vsc

*** 001_list.sh ***

> module list

>>> 001_list.sh: PASS

*** 002_avail.sh ***

> module avail
> module avail GCC
> module avail GCC/4.9.3

>>> 002_avail.sh: PASS

*** 003_load.sh ***

> module load GCC
> module load GCC/4.9.3
> module load intel
> module load foss
> module load Python/2.7.11-intel-2016a
> module load GCC/4.9.3-2.25 OpenMPI/1.10.2-GCC-4.9.3-2.25 OpenBLAS/0.2.15-GCC-4.9.3-2.25-LAPACK-3.6.0 FFTW/3.3.4-gompi-2016a

>>> 003_load.sh: PASS

*** 004_purge.sh ***

> module load Python/2.7.11-intel-2016a
> module purge
> module load cluster
> module purge -force
> module load cluster
> module purge -force
> module load cluster/swalot

>>> 004_purge.sh: PASS

*** 005_swap.sh ***

> module load GCC/4.9.3
> module swap GCC/5.3.0
> module swap GCC GCC/4.9.3
> module swap GCC/4.9.3 GCC/5.3.0

>>> 005_swap.sh: PASS

*** 006_unload.sh ***

> module load GCC/4.9.3
> module unload GCC
> module load GCC/4.9.3
> module unload GCC/4.9.3

>>> 006_unload.sh: PASS

*** 007_spider.sh ***

> module spider intel
> module spider intel/2016a
> module --show-hidden spider intel/2016a

>>> 007_spider.sh: PASS

*** 010_stdout_stderr.sh ***

> module list
> module avail

>>> 010_stdout_stderr.sh: PASS

*** 050_ml.sh ***

> ml av GCC/4.9.3
> ml
> ml GCC/4.9.3
> ml
> ml -GCC/4.9.3
> ml

>>> 050_ml.sh: PASS

*** 051_collections.sh ***

> ml foss/2016a
> ml Python/2.7.11-intel-2016a
> ml save this_is_just_a_test_collection_for_module_integration_test_051
> ml describe this_is_just_a_test_collection_for_module_integration_test_051
> ml purge
> ml restore this_is_just_a_test_collection_for_module_integration_test_051
> ml purge
> module swap cluster/swalot cluster/delcatty
> ml purge
> ml restore this_is_just_a_test_collection_for_module_integration_test_051
> ml purge

>>> 051_collections.sh: PASS

*** 100_lmod_cache.sh ***

> module avail  # 2s time limit

>>> 100_lmod_cache.sh: PASS

*** 101_LD_LIBRARY_PATH.sh ***

> module load GCC/4.9.3-2.25
> module load OpenMPI/1.10.2-GCC-4.9.3-2.25
> checking $LD_LIBRARY_PATH...

>>> 101_LD_LIBRARY_PATH.sh: PASS

*** 102_symlink_modulepath.sh ***

> module use /tmp/vsc40023/0B810Q/symlinked_modules
> module avail test/1.2.3

>>> 102_symlink_modulepath.sh: PASS

*** 103_tcl2lua_LD_PRELOAD.sh ***

> module load jemalloc
> module show jemalloc
> ml intel/2016a && LD_LIBRARY_PATH='' LD_PRELOAD='' ml MariaDB/10.1.14-intel-2016a

>>> 103_tcl2lua_LD_PRELOAD.sh: PASS

TEST RESULT: all 14 passed!
```

File: SitePackage.lua [hunk 1]
```diff
@@ -124,38 +125,53 @@ local function startup_hook(usrCmd)
...
+    -- kind is either lmoderror, lmodwarning or lmodmessage
+    -- key is a unique key for the message (see messageT.lua)
+    -- msg is the actual message to display (as string)
+    -- t is a table with the keys used in msg
+    dbg.start{"errwarnmsg_hook"}
 
-    dbg.start{"msg_hook"}
+    dbg.print{"kind: ", kind}
+    dbg.print{"Msg: ", msg}
+    dbg.print{"keys: ", t}
```
Comment (wpoely86) [inline SitePackage.lua]: these were purly for debugging, you can drop them

File: buildrpmfromspec.sh [hunk 2]
```diff
@@ -62,6 +62,7 @@ wget -O $lmod_src "https://github.com/TACC/Lmod/archive/${vers}.tar.gz"
 mv $lmod_src $src_dir
 
 cp -a *.patch admin.list SitePackage.lua macros.Lmod run_lmod_cache.py $src_dir
+#cp -a admin.list SitePackage.lua macros.Lmod run_lmod_cache.py $src_dir
```
Comment (wpoely86) [inline buildrpmfromspec.sh]: drop?

File: SitePackage.lua [hunk 3]
```diff
@@ -124,38 +125,53 @@ local function startup_hook(usrCmd)
...
-        local mStack = ModuleStack:moduleStack()
+        local sname = t.sn
+        local frameStk = FrameStk:singleton()
 
         local errmsg = {"A different version of the '"..sname.."' module is already loaded (see output of 'ml')."}
-        if not mStack:empty() then
-            errmsg[#errmsg+1] = "You should load another '"..mStack:sn().."' module for that is compatible with the currently loaded version of '"..sname.."'."
-            errmsg[#errmsg+1] = "Use 'ml spider "..mStack:sn().."' to get an overview of the available versions."
+        if not frameStk:empty() then
+            local compat_msg = "' module for that is compatible with the currently loaded version of '"
```
Comment (wpoely86) [inline SitePackage.lua:142]: why put this in a different variable?

**Review by wpoely86 — APPROVED** [review]

File: SitePackage.lua (hunk 3 above)
Comment (boegel) [reply SitePackage.lua:142]: because otherwise the line is too long, and then `luacheck` spits out a warning that makes Jenkins block the PR... :)

**Pull request merged**